### ISSUE TITLE
Poor result ranking for workspaceSymbol

### ISSUE THREAD

**Issue by HighCommander4** [clangd/vscode-clangd#81]
If I'm editing the LLVM codebase with vscode + clangd and I perform a `workspaceSymbol` search for `TypeDecl`, this is what I get:

![workspacesymbol-screenshot](https://user-images.githubusercontent.com/1751085/95007242-7c850080-05db-11eb-81b6-d2e8cccb0257.png)

Note that `clang::TypeDecl` is the very last result, and all sorts of identifiers where the leaf segment is only a partial as opposed to an exact match are ranked higher than it.

I believe this is clearly suboptimal. Symbols where the leaf segment of the qualified name is an exact match should be ranked higher than ones where it's a partial match.

**Comment by sam-mccall** [comment]
Yeah, I'm happy to add a `score` extension to our responses, but there's not that much we can do here, this is a bug somewhere between VSCode and LSP.
(LSP doesn't let us communicate ordering beyond providing an order, and VSCode ignores the order we provide).

Any ideas before we close this?

**Comment by Trass3r** [comment]
The question is where the order is lost. Can barely find the relevant code in vscode, only [this](https://github.com/microsoft/vscode/blob/c185f7ab939ca3007c62039dd70ec01862286b93/src/vs/workbench/api/common/extHostApiCommands.ts#L193)

**Comment by HighCommander4** [comment]
It looks like the culprit is our fix for #31.

What VSCode appears to be doing is:

  * compare `symbol.name` to the query string to discriminate between exact (prefix) matches and partial matches
  * rank exact matches above partial matches
  * within the group of partial matches, sort them alphabetically

(I'm basing this on experimentation, I haven't tracked down the relevant VSCode source.)

So, if `symbol.name` contains the leaf name only, and `symbol.containerName` contains the qualifier, as the API intends, then all is well, and exact matches get ranked above partial matches.

This is indeed how the server provides the symbols. However, we [added middleware code to the client](https://github.com/clangd/vscode-clangd/blob/11f5b680647de3c195997515ecc85d0983fe4395/src/clangd-context.ts#L117) to manipulate the server results in the following way:

```ts
        provideWorkspaceSymbols: async (query, token, next) => {
          let symbols = await next(query, token);
          return symbols.map(symbol => {
            if (symbol.containerName)
              symbol.name = `${symbol.containerName}::${symbol.name}`;
            // Always clean the containerName to avoid displaying it twice.
            symbol.containerName = '';
            return symbol;
          })
        },
```

By including the qualifier in `symbol.name`, every match becomes a partial match (for names with qualifiers where the full qualifier is not included in the query), and we just get an alphabetical ordering of all the matches. So, by doing this, we solved one problem (the one in #31, that qualified names could not be used as a search query) but created another (this one, the poor ordering).

**Comment by HighCommander4** [comment]
I proposed a fix which limits the manipulation added for #31, to just the cases where the query string is in fact qualified. That should solve the problem for the vast majority of cases. (Technically, the problem can still occur for partially-qualified query strings, but those are much less likely to have a mixture of exact and partial matches.)
